packages/core/src/application/application_ref.ts

```diff
@@ -583,10 +583,20 @@ export class ApplicationRef {
         }
         this.detectChangesInView(_lView, notifyErrorHandler, isFirstPass);
       }
-      this.afterRenderEffectManager.execute();
-
       runs++;
-    } while (this._views.some(({_lView}) => shouldRecheckView(_lView)));
+
+      this.afterRenderEffectManager.executeInternalCallbacks();
+      // If we have a newly dirty view after running internal callbacks, recheck the views again before running user-provided callbacks
+      if (this._views.some(({_lView}) => shouldRecheckView(_lView))) {
+        continue;
+      }
+
+      this.afterRenderEffectManager.execute();
```

Comment:
nit: we can extract `this.afterRenderEffectManager` to a const outside a loop and use it inside:
```suggestion
      afterRenderEffectManager.execute();
```
(less property reads + better minification)